src/scene/forward-renderer.js

```diff
@@ -2290,59 +2293,55 @@ Object.assign(ForwardRenderer.prototype, {
         // #endif
     },
 
-    updateShaders: function (drawCalls) {
-        // #ifdef PROFILER
-        var time = now();
-        // #endif
+    // funtion returns unique set of materials from drawCalls
+    // material parameter is expected to be a Set to allow for fast duplicates removal
+    getUniqueMaterials: function (materials, drawCalls) {
 
-        var i;
-        // Collect materials
-        var materials = [];
-        for (i = 0; i < drawCalls.length; i++) {
-            var drawCall = drawCalls[i];
-            if (drawCall.material !== undefined) {
-                if (materials.indexOf(drawCall.material) === -1) {
-                    materials.push(drawCall.material);
-                }
+        var mat;
+        for (var i = 0; i < drawCalls.length; i++) {
+            mat = drawCalls[i].material;
+            if (mat) {
+                materials.add(mat);
             }
         }
+
+        return materials;
+    },
+
+    updateShaders: function (drawCalls) {
+
         // Clear material shaders
-        for (i = 0; i < materials.length; i++) {
-            var mat = materials[i];
+        this.getUniqueMaterials(_tempMaterialSet, drawCalls);
+        for (var mat of _tempMaterialSet) {
```

Comment:
There are two loops here with extra function call. Why not iterate once through draw calls with track of updated materials using a set (as is now), so eliminating for second iteration through set of materials.

Reply:
a fine suggestion. I've done that now.